File: src/disk_tree/diff.py
"""Best-first pruned recursive diff between two scans (spec: diff-and-search.md §3a).

Deltas propagate up the tree, so a change 9 levels deep shows at depth 1 — as an
undifferentiated Δ on one child. This module does the drill-down server-side, in
one pass, and only down *changed spines*:

- `added`/`removed` dirs are never descended — their aggregate row already tells
  the whole story (layer-2 dirs carry `size`/`n_desc`).
- Common dirs whose `(size, n_desc, n_children, mtime)` all match are pruned.
- Same size & count but a different mtime → `touched` (emitted; dirs are
  still descended, since a net-zero rename hides below).
  Heuristic: compensating changes can hide (the eventual `digest` column makes
  this exact — spec §3d). `mtime` is in the descend-trigger so a same-size
  rename (net-zero Δ) is still found and surfaces as `added` + `removed` rows.
- Expansion order is best-first on `|Δsize|` (not BFS), so a budget cuts the
  walk where the *least* signal is; unexpanded frontier dirs are marked
  `pruned` so callers know change may hide below them.

Also home to `resolve_blob` / `resolve_chunk_for_path` (moved from `server.py`
so the CLI can share them without importing Flask).
"""

from __future__ import annotations

import heapq
from dataclasses import dataclass, field
from functools import lru_cache
from os.path import exists, getmtime, isabs, join
from typing import Callable, Protocol

import pandas as pd
import pyarrow.parquet as pq

from . import config as _config


def resolve_blob(blob_ref: str) -> str:
    """Resolve a parquet blob ref to its absolute path.

    Honors legacy absolute refs and ignores DuckDB/SQLite opaque refs.
    Reads SCANS_DIR via the config module so tests can monkeypatch it.
    """
    if not blob_ref or blob_ref.startswith(('ddb:', 'sqlite:')):
        return blob_ref
    return blob_ref if isabs(blob_ref) else join(_config.SCANS_DIR, blob_ref)


def _chunk_map(parquet_path: str) -> dict[str, str] | None:
    """path → child_scan_id for the chunk-pointer rows of a hybrid parquet
    (None when the blob has no `child_scan_id` column). Keyed on mtime because
    delete updates rewrite blobs in place."""
    return _chunk_map_cached(parquet_path, getmtime(parquet_path))


@lru_cache(maxsize=64)
def _chunk_map_cached(parquet_path: str, mtime: float) -> dict[str, str] | None:
    if 'child_scan_id' not in pq.read_schema(parquet_path).names:
        return None
    df = pd.read_parquet(parquet_path, columns=['path', 'child_scan_id'])
    df = df[df['child_scan_id'].notna()]
    return dict(zip(df['path'], df['child_scan_id']))


def resolve_chunk_for_path(blob_ref: str, rel_path: str) -> tuple[str, str]:
    """Resolve the actual blob_ref and rebased rel_path for a path that may be in a chunk.

    If rel_path maps to a chunked subtree, returns (chunk_blob_ref, rebased_path).
    Otherwise returns (blob_ref, rel_path) unchanged.

    Note: Only hybrid backend uses chunked parquets. DuckDB/SQLite blob refs
    (prefixed with 'ddb:' or 'sqlite:') don't have chunks.
    """
    if not rel_path or rel_path == '.':
        return blob_ref, rel_path

    # DuckDB and SQLite backends don't use chunking
    if blob_ref.startswith('ddb:') or blob_ref.startswith('sqlite:'):
        return blob_ref, rel_path

    chunks = _chunk_map(resolve_blob(blob_ref))
    if chunks is None:
        return blob_ref, rel_path

    # Check if any ancestor of rel_path has a child_scan_id
    parts = rel_path.split('/')
    for i in range(len(parts)):
        chunk_ref = chunks.get('/'.join(parts[:i+1]))
        if chunk_ref is not None and exists(resolve_blob(chunk_ref)):
            # Rebase the remaining path relative to chunk root
            remaining = '/'.join(parts[i+1:]) if i + 1 < len(parts) else '.'
            # Recursively resolve in case of nested chunks
            return resolve_chunk_for_path(chunk_ref, remaining)

    return blob_ref, rel_path


class LoadFn(Protocol):
    def __call__(
        self,
        blob_ref: str,
        max_depth: int | None = None,
        min_depth: int | None = None,
        follow_refs: bool = False,
        path_prefix: str | None = None,
    ) -> pd.DataFrame: ...


class ScanSource:
    """Per-scan children loader for the diff walk.

    Takes paths *relative to the compared uri* (`''` = the uri itself), rebases
    into the scan's own coordinates (the scan may be of an ancestor path),
    resolves hybrid chunk boundaries, and loads exactly one directory's
    children per call — cheap via depth + path-prefix pushdown.
    """

    def __init__(
        self,
        blob: str,
        scan_path: str,
        uri: str,
        load: LoadFn,
        resolve: Callable[[str, str], tuple[str, str]] | None = None,
    ):
        """`resolve` maps (blob, rel_path) across hybrid chunk boundaries —
        pass `resolve_chunk_for_path` for hybrid-backed scans (it resolves refs
        via the configured SCANS_DIR, so it's not the default: backends with
        their own scans_dir, and chunk-free backends, want identity)."""
        self.blob = blob
        self.load = load
        self.resolve = resolve or (lambda b, p: (b, p))
        if scan_path == uri:
            self.rel_prefix = ''
        else:
            self.rel_prefix = uri[len(scan_path):].lstrip('/')

    def children(self, rel: str) -> pd.DataFrame:
        """Direct children of `rel`, indexed by child name.

        Rows at exactly one depth under one prefix *are* the direct children —
        no parent-column mask needed.
        """
        if self.rel_prefix and rel:
            full = f"{self.rel_prefix}/{rel}"
        else:
            full = self.rel_prefix or rel
        blob, rebased = self.resolve(self.blob, full or '.')
        at_root = rebased in ('', '.')
        d = 0 if at_root else rebased.count('/') + 1
        df = self.load(
            blob,
            min_depth=d + 1,
            max_depth=d + 1,
            path_prefix=None if at_root else rebased,
        )
        if not df.empty:
            # Depth pushdown is best-effort (blobs without a `depth` column
            # load unfiltered) — enforce "exactly one level below `rebased`"
            # here so correctness never depends on the backend's fidelity.
            depths = df['path'].map(lambda p: 0 if p == '.' else p.count('/') + 1)
            mask = depths == d + 1
            if not at_root:
                mask &= df['path'].str.startswith(rebased + '/')
            df = df[mask]
        if df.empty:
            return df.set_index(df['path'] if 'path' in df.columns else pd.Index([]))
        names = df['path'] if at_root else df['path'].str.rsplit('/', n=1).str[-1]
        df = df.set_index(names)
        return df[~df.index.duplicated()]


def _neq(a, b) -> bool:
    """NaN-safe inequality (both-NaN counts as equal, unlike `!=`)."""
    a_na, b_na = pd.isna(a), pd.isna(b)
    if a_na and b_na:
        return False
    if a_na or b_na:
        return True
    return bool(a != b)


@dataclass
class DeltaRow:
    path: str      # relative to the compared uri
    depth: int     # levels below the compared uri (1 = direct child)
    kind: str
    status: str    # added | removed | changed | touched | unchanged
                   # touched: same size & n_desc, mtime differs (a rename /
                   # net-zero churn / plain `touch` — bytes didn't move)
    size_a: int
    size_b: int
    n_desc_a: int
    n_desc_b: int
    expanded: bool = False   # a dir we descended into
    pruned: bool = False     # a dir with differing stats we did NOT descend into (budget/depth)

    @property
    def size_delta(self) -> int:
        return self.size_b - self.size_a

    @property
    def n_desc_delta(self) -> int:
        return self.n_desc_b - self.n_desc_a


@dataclass
class UnchangedRest:
    """Unchanged children of one expanded dir that were *not* emitted (beyond
    the top-K): enough to label the grey remainder without shipping rows."""
    count: int      # direct children
    size: int       # Σ size (both sides equal — they're unchanged)
    n_desc: int     # Σ n_desc — descendants below those children


@dataclass
class RecursiveDiffResult:
    rows: list[DeltaRow]   # sorted by |Δsize| desc
    expansions: int        # directories whose children were loaded (root included)
    truncated: bool        # budget or max_depth cut the walk short
    # The biggest unchanged siblings seen while expanding each dir (labeled
    # grey context for the treemap), and the aggregate of the rest, keyed by
    # parent rel path ('' = the compared uri). Empty when `include_unchanged`
    # already put every unchanged row in `rows`.
    unchanged_top: list[DeltaRow] = field(default_factory=list)
    unchanged_rest: dict[str, UnchangedRest] = field(default_factory=dict)


def _ival(row, col: str) -> int:
    if row is None:
        return 0
    v = row.get(col, 0)
    return 0 if pd.isna(v) else int(v)


def recursive_diff(
    src_a: ScanSource,
    src_b: ScanSource,
    budget: int = 100,
    max_depth: int | None = None,
    include_unchanged: bool = False,
    unchanged_top: int = 8,
) -> RecursiveDiffResult:
    """Walk changed spines best-first; return the delta frontier.

    `unchanged_top`: per expanded dir, keep this many of the biggest unchanged
    children (by size) in `unchanged_top`, summarizing the rest in
    `unchanged_rest[parent]` — free, since every child was compared anyway.
    """
    rows: list[DeltaRow] = []
    top_rows: list[DeltaRow] = []
    rest: dict[str, UnchangedRest] = {}
    emitted: set[str] = set()
    by_path: dict[str, DeltaRow] = {}
    # heap entries: (-|Δsize|, depth, seq, rel) — seq breaks ties FIFO
    heap: list[tuple[int, int, int, str]] = [(0, 0, 0, '')]
    seq = 0
    expansions = 0
    truncated = False

    while heap:
        if expansions >= budget:
            truncated = True
            break
        _, d, _, rel = heapq.heappop(heap)
        ca = src_a.children(rel)
        cb = src_b.children(rel)
        expansions += 1
        if rel:
            by_path[rel].expanded = True

        unchanged_here: list[DeltaRow] = []
        for name in sorted(set(ca.index) | set(cb.index)):
            child_rel = f"{rel}/{name}" if rel else name
            ra = ca.loc[name] if name in ca.index else None
            rb = cb.loc[name] if name in cb.index else None
            kind_a = ra['kind'] if ra is not None else None
            kind_b = rb['kind'] if rb is not None else None
            row = DeltaRow(
                path=child_rel,
                depth=d + 1,
                kind=kind_b or kind_a or '',
                status='',
                size_a=_ival(ra, 'size'),
                size_b=_ival(rb, 'size'),
                n_desc_a=_ival(ra, 'n_desc'),
                n_desc_b=_ival(rb, 'n_desc'),
            )
            descend = False
            if ra is None:
                row.status = 'added'
            elif rb is None:
                row.status = 'removed'
            elif kind_a != kind_b:
                # dir↔file swap: incomparable subtrees, report don't descend
                row.status = 'changed'
            else:
                if row.size_delta or row.n_desc_delta:
                    row.status = 'changed'
                elif _neq(ra.get('mtime'), rb.get('mtime')):
                    row.status = 'touched'
                else:
                    row.status = 'unchanged'
                if kind_a == 'dir':
                    # mtime is in the trigger so net-zero renames are still found
                    descend = any(
                        _neq(ra.get(c), rb.get(c))
                        for c in ('size', 'n_desc', 'n_children', 'mtime')
                    )

            by_path[child_rel] = row
            if descend:
                if max_depth is not None and d + 1 >= max_depth:
                    row.pruned = True
                    truncated = True
                else:
                    seq += 1
                    heapq.heappush(heap, (-abs(row.size_delta), d + 1, seq, child_rel))
            if row.status != 'unchanged' or include_unchanged or row.pruned:
                rows.append(row)
                emitted.add(child_rel)
            elif not row.pruned:
                unchanged_here.append(row)
        if unchanged_here:
            unchanged_here.sort(key=lambda r: (-r.size_b, r.path))
            top_rows.extend(unchanged_here[:unchanged_top])
            tail = unchanged_here[unchanged_top:]
            if tail:
                rest[rel] = UnchangedRest(
                    count=len(tail),
                    size=sum(r.size_b for r in tail),
                    n_desc=sum(r.n_desc_b for r in tail),
                )

    # Anything still queued was never expanded: mark it (and surface it even if
    # its own Δ is zero — differing stats below mean unexplored change).
    for _, _, _, rel in heap:
        row = by_path.get(rel)
        if row is not None:
            row.pruned = True
            truncated = True
            if rel not in emitted:
                rows.append(row)
                emitted.add(rel)

    rows.sort(key=lambda r: (-abs(r.size_delta), r.path))
    return RecursiveDiffResult(
        rows=rows, expansions=expansions, truncated=truncated,
        unchanged_top=top_rows, unchanged_rest=rest,
    )

File: ui/cfn/diff.ts
/** On-the-fly scan-to-scan diff for the static compare Function — the
 *  serverless analog of the Flask `/api/compare` walk (spec
 *  `specs/public-diff-demo.md` Phase 3). Given two scans' rows *rebased to the
 *  compared uri* (path `.` = the uri, `child` = depth 1, `child/gc` = depth 2),
 *  it aligns them per path and emits the flat one-level child diff (the table)
 *  and a depth-≤2 frontier (the treemap): direct children, plus one level into
 *  each *changed* dir, with deeper-changed dirs marked `pruned` so the widget
 *  offers a drill (`DiffTreemap.fetchSubtree`). Undrawable rows (bytes and |Δ|
 *  both under the viewport byte-floor) are dropped. Statuses match
 *  `diff_index.py`: added | removed | changed | touched | unchanged. */
import type { TreeRow } from './parquet'

export type DiffStatus = 'added' | 'removed' | 'changed' | 'touched' | 'unchanged'

/** One path's status across the two scans (`diff_index.py` vectorized rules):
 *  present on one side only → added/removed; both with differing size / count /
 *  kind → changed; both equal but mtime moved → touched; else unchanged. */
export function diffStatus(a: TreeRow | undefined, b: TreeRow | undefined): DiffStatus {
  if (!a) return 'added'
  if (!b) return 'removed'
  if (a.size !== b.size || a.n_desc !== b.n_desc || a.kind !== b.kind) return 'changed'
  if (a.mtime !== b.mtime) return 'touched'
  return 'unchanged'
}

/** A row as `/api/compare` (non-recursive) returns it — the table shape. */
export interface CompareRow {
  path: string
  uri: string
  kind: 'file' | 'dir'
  parent: string | null
  status: DiffStatus
  size: number | null
  size_old?: number
  mtime: number | null
  n_desc: number | null
  n_children: number | null
  size_delta: number
  n_desc_old?: number
  n_desc_delta?: number
}

/** A row as `/api/compare?recursive=1` returns it — the treemap frontier. */
export interface CompareRecRow {
  path: string
  uri: string
  depth: number
  kind: 'file' | 'dir'
  status: DiffStatus
  size_a: number
  size_b: number
  size_delta: number
  n_desc_a: number
  n_desc_b: number
  n_desc_delta: number
  expanded: boolean
  pruned: boolean
}

export interface DiffSummary {
  added: number
  removed: number
  changed: number
  unchanged: number
  total_delta: number
}

/** A scan row rebased to the compared uri (`scan.ts` `toApi` output). */
export interface RebasedRow extends TreeRow { uri: string }

const size = (r: RebasedRow | undefined): number => r?.size ?? 0
const ndesc = (r: RebasedRow | undefined): number => r?.n_desc ?? 0

/** Index rebased rows by their uri-relative path. */
function byPath(rows: RebasedRow[]): Map<string, RebasedRow> {
  const m = new Map<string, RebasedRow>()
  for (const r of rows) m.set(r.path, r)
  return m
}

/** Direct children (relative depth 1) present in either side, by path. */
function childrenAt(
  parentRel: string,
  a: Map<string, RebasedRow>,
  b: Map<string, RebasedRow>,
): string[] {
  const want = parentRel === '.' ? 1 : parentRel.split('/').length + 1
  const under = (p: string) => (parentRel === '.' ? !p.includes('/') && p !== '.' : p.startsWith(parentRel + '/') && p.split('/').length === want)
  const out = new Set<string>()
  for (const m of [a, b]) for (const [p, r] of m) if (r.depth === want && under(p)) out.add(p)
  return [...out]
}

export interface DiffOptions {
  /** Absolute uri of the compared root (for `rowUri`). */
  uri: string
  /** `minFrac`: drop rows whose max bytes and |Δ| are both under
   *  `minFrac · max(rootA, rootB)` — undrawable cells. 0 keeps everything. */
  minFrac?: number
  /** Max frontier rows returned (best-first by |Δ|), default 2000. */
  maxRows?: number
}

const rowUri = (uri: string, rel: string): string =>
  rel === '.' ? uri : `${uri}/${rel}`

/** Absolute uri for a uri-relative child path, preferring a row's own `uri`. */
const childUri = (uri: string, rel: string, r?: RebasedRow): string =>
  r?.uri ?? rowUri(uri, rel)

/** Per-status counters (`touched` tracked internally but not in the response
 *  summary, which mirrors Flask's added/removed/changed/unchanged). */
type Counts = Record<DiffStatus, number>
const zeroCounts = (): Counts => ({ added: 0, removed: 0, changed: 0, touched: 0, unchanged: 0 })
const toSummary = (c: Counts, total_delta: number): DiffSummary =>
  ({ added: c.added, removed: c.removed, changed: c.changed, unchanged: c.unchanged, total_delta })

/** The flat one-level child diff (the table): every direct child of the uri. */
export function flatDiff(aRows: RebasedRow[], bRows: RebasedRow[], o: DiffOptions): { rows: CompareRow[]; summary: DiffSummary } {
  const a = byPath(aRows), b = byPath(bRows)
  const rows: CompareRow[] = []
  const counts = zeroCounts()
  let total = 0
  for (const rel of childrenAt('.', a, b)) {
    const ra = a.get(rel), rb = b.get(rel)
    const status = diffStatus(ra, rb)
    const delta = size(rb) - size(ra)
    counts[status] += 1
    total += delta
    if (status === 'unchanged') continue
    const r = (rb ?? ra)!
    rows.push({
      path: rel,
      uri: childUri(o.uri, rel, rb ?? ra),
      kind: r.kind,
      parent: '.',
      status,
      size: rb ? rb.size : null,
      size_old: ra ? size(ra) : undefined,
      mtime: r.mtime,
      n_desc: rb ? rb.n_desc : null,
      n_children: r.n_children,
      size_delta: delta,
      n_desc_old: ra ? ndesc(ra) : undefined,
      n_desc_delta: ndesc(rb) - ndesc(ra),
    })
  }
  rows.sort((x, y) => Math.abs(y.size_delta) - Math.abs(x.size_delta))
  return { rows, summary: toSummary(counts, total) }
}

/** The depth-≤2 frontier (the treemap) + labeled unchanged context. */
export function recursiveDiff(aRows: RebasedRow[], bRows: RebasedRow[], o: DiffOptions): {
  rows: CompareRecRow[]
  unchanged: { top: CompareRecRow[]; rest: Record<string, { count: number; size: number; n_desc: number }> }
  summary: DiffSummary & { expansions: number; truncated: boolean }
} {
  const a = byPath(aRows), b = byPath(bRows)
  const rootA = a.get('.'), rootB = b.get('.')
  const floor = (o.minFrac ?? 0) * Math.max(size(rootA), size(rootB))
  const drawable = (ra?: RebasedRow, rb?: RebasedRow): boolean => {
    const bytes = Math.max(size(ra), size(rb))
    return bytes >= floor || Math.abs(size(rb) - size(ra)) >= floor
  }

  const rows: CompareRecRow[] = []
  const unchangedTop: CompareRecRow[] = []
  const rest: Record<string, { count: number; size: number; n_desc: number }> = {}
  const counts = zeroCounts()
  let total = 0
  let expansions = 0
  let truncated = false

  const mkRow = (rel: string, depth: number, ra: RebasedRow | undefined, rb: RebasedRow | undefined, expanded: boolean, pruned: boolean): CompareRecRow => ({
    path: rel,
    uri: childUri(o.uri, rel, rb ?? ra),
    depth,
    kind: (rb ?? ra)!.kind,
    status: diffStatus(ra, rb),
    size_a: size(ra), size_b: size(rb), size_delta: size(rb) - size(ra),
    n_desc_a: ndesc(ra), n_desc_b: ndesc(rb), n_desc_delta: ndesc(rb) - ndesc(ra),
    expanded, pruned,
  })

  // Keep a dir's biggest unchanged children as labeled grey context, aggregate
  // the rest into one row (`key` `''` = the compared uri itself).
  const finalizeUnchanged = (key: string, tops: CompareRecRow[]) => {
    tops.sort((x, y) => Math.max(y.size_a, y.size_b) - Math.max(x.size_a, x.size_b))
    for (const t of tops.slice(0, 4)) unchangedTop.push(t)
    const agg = { count: 0, size: 0, n_desc: 0 }
    for (const t of tops.slice(4)) { agg.count += 1; agg.size += t.size_b; agg.n_desc += t.n_desc_b }
    if (agg.count) rest[key] = agg
  }

  // Depth 1: every changed direct child (best-first frontier). A changed dir is
  // expanded one level (depth 2). Unchanged children of the compared uri and of
  // each expanded dir become that dir's grey context.
  const rootTops: CompareRecRow[] = []
  for (const rel of childrenAt('.', a, b)) {
    const ra = a.get(rel), rb = b.get(rel)
    const status = diffStatus(ra, rb)
    if (status === 'unchanged') { counts.unchanged += 1; if (drawable(ra, rb)) rootTops.push(mkRow(rel, 1, ra, rb, false, false)); continue }
    counts[status] += 1; total += size(rb) - size(ra)
    if (!drawable(ra, rb)) continue
    const isDir = (rb ?? ra)!.kind === 'dir'
    // Descend only into a *changed* dir present on both sides (added/removed
    // subtrees are a single row by design).
    const expand = isDir && status === 'changed'
    rows.push(mkRow(rel, 1, ra, rb, expand, false))
    if (!expand) continue
    expansions += 1
    const tops: CompareRecRow[] = []
    for (const c of childrenAt(rel, a, b)) {
      const ca = a.get(c), cb = b.get(c)
      const cStatus = diffStatus(ca, cb)
      if (cStatus === 'unchanged') { if (drawable(ca, cb)) tops.push(mkRow(c, 2, ca, cb, false, false)); continue }
      if (!drawable(ca, cb)) continue
      // A changed dir at depth 2 isn't descended → mark pruned so the widget
      // offers a drill (`fetchable`: changed dir + non-zero Δ).
      const cIsDir = (cb ?? ca)!.kind === 'dir'
      rows.push(mkRow(c, 2, ca, cb, false, cIsDir && cStatus === 'changed'))
    }
    finalizeUnchanged(rel, tops)
  }
  finalizeUnchanged('', rootTops)

  rows.sort((x, y) => Math.abs(y.size_delta) - Math.abs(x.size_delta))
  const cap = o.maxRows ?? 2000
  if (rows.length > cap) { truncated = true; rows.length = cap }
  return { rows, unchanged: { top: unchangedTop, rest }, summary: { ...toSummary(counts, total), expansions, truncated } }
}
